### PR TITLE
Work system updates

### PR DESCRIPTION
**Pull request by darinkrauss**
- Resolve overflow issue with exponential fallback duration calculations
- Fail if registering work processor factories after coordinator start
- Use crypto random for duration jitter
- Pass time to ProcessResult builder functions
- Prevent unnecessarily deep work context due to fields
- Allow override of Now time for testing
- Add data work general functionality
- Update tests

### CODE REVIEW

File: work/base/process_result.go [hunk 1]
```diff
@@ -43,62 +44,100 @@ func (p *ProcessResultBuilder) Failing(ctx context.Context, wrk *work.Work, err
...
-	return now.Add(duration)
+	return tm.Add(duration)
 }
+
+func durationWithJitterExponential(duration time.Duration, durationJitter time.Duration, exponent int) time.Duration {
+	if duration < 0 || exponent < 0 {
+		return 0
+	}
+	duration = durationExponential(duration, exponent)
+	durationJitter = time.Duration(crypto.RandomInt64N(int64(durationExponential(durationAbsolute(durationJitter), exponent))))
```
[the remark this answers is not in the document]
Comment (darinkrauss) [reply work/base/process_result.go:114]: Good catch. Also explicitly added test with duration jitter of zero.

[the remark this answers is not in the document]
Comment (darinkrauss) [reply work/base/process_result.go:114]: Fix will be in future PR.

File: work/base/process_result.go [hunk 2]
```diff
@@ -43,62 +44,100 @@ func (p *ProcessResultBuilder) Failing(ctx context.Context, wrk *work.Work, err
...
+
+func durationExponential(duration time.Duration, exponent int) time.Duration {
+	if duration <= 0 || exponent < 0 {
+		return 0
+	} else if exponent == 0 {
+		return duration
+	} else if exponent > int(math.Log2(float64(durationMaximum)/float64(duration))) {
+		return durationMaximum
+	} else {
+		return time.Duration(int64(duration) * (1 << exponent))
```
[the remark this answers is not in the document]
Comment (darinkrauss) [reply work/base/process_result.go:131]: Very subtle bug. Good one. Fix in later PR.

File: work/base/processor.go [hunk 3]
```diff
@@ -78,15 +84,21 @@ func (p *Processor[W]) WorkClient() work.Client {
...
+	p.AddFieldsToContext(log.Fields{key: value})
 }
 
 func (p *Processor[W]) AddFieldsToContext(fields log.Fields) {
-	p.context = log.ContextWithFields(p.context, fields)
+	if p.contextFields == nil {
+		p.contextFields = log.Fields{}
+	}
+	maps.Copy(p.contextFields, fields)
 }
```
[the remark this answers is not in the document]
Comment (darinkrauss) [reply work/base/processor.go:102]: Processor is designed to only be used once.

File: work/service/coordinator.go [hunk 4]
```diff
@@ -96,6 +99,10 @@ func (c *Coordinator) RegisterProcessorFactory(processorFactory work.ProcessorFa
 		return errors.New("processor factory is missing")
 	}
 
+	if c.workersCompletionChannel != nil {
+		return errors.New("coordinator already started")
+	}
+
```
[the remark this answers is not in the document]
Comment (darinkrauss) [reply work/service/coordinator.go:105]: Doh, true! Fix in later PR.

File: work/service/coordinator.go [hunk 5]
```diff
@@ -350,7 +357,7 @@ func (c *Coordinator) completeWork(completion *coordinatorProcessingCompletion)
 
 func (c *Coordinator) startTimer() {
 	jitter := int64(float64(c.frequency) * CoordinatorDelayJitter)
-	frequencyWithJitter := c.frequency + time.Duration(rand.Int63n(jitter*2+1)-jitter)
+	frequencyWithJitter := c.frequency + time.Duration(crypto.RandomInt64N(jitter*2+1)-jitter)
 	if c.timer == nil {
```
[the remark this answers is not in the document]
Comment (darinkrauss) [reply work/service/coordinator.go:361]: Not an issue. Other code ensures jitter to be between 1 and 3*10^11 (5 minutes) and cannot yield negative, zero, or overflow.

File: work/base/processor.go [hunk 6]
```diff
@@ -78,15 +84,21 @@ func (p *Processor[W]) WorkClient() work.Client {
...
 
 func (p *Processor[W]) Context() context.Context {
+	if len(p.contextFields) > 0 {
+		return log.ContextWithFields(p.context, p.contextFields)
+	}
 	return p.context
```
[the remark this answers is not in the document]
Comment (darinkrauss) [reply work/base/processor.go:90]: Fix in later PR.

**Review by toddkazakov — APPROVED** [review]
